### PR TITLE
CI: add single test for powered-off CirrOS VM

### PR DESCRIPTION
**Pull request by mrnold**
## Why the changes were made

Add a test for backup/restore on powered-off VMs. This is the first step in addressing [OADP-4231](https://issues.redhat.com/browse/OADP-4231).

## How to test the changes made

Run virtualization E2E tests with: make TEST_VIRT=true test-e2e

### CODE REVIEW

File: tests/e2e/backup_restore_suite_test.go [hunk 1]
```diff
@@ -227,7 +227,7 @@ func runBackup(brCase BackupRestoreCase, backupName string) bool {
 	return nsRequiresResticDCWorkaround
 }
 
-func runRestore(brCase BackupRestoreCase, backupName, restoreName string, nsRequiresResticDCWorkaround bool) {
+func runRestore(brCase BackupRestoreCase, backupName, restoreName string, nsRequiresResticDCWorkaround bool, expectedErr error) {
```
Comment (mateusoliveira43) [inline tests/e2e/backup_restore_suite_test.go:230]: to avoid changes in backup_restore_suite_test.go and in must-gather_suite_test.go, I would copy this function to virt_backup_restore_suite_test.go with a TODO to be removed after problem is fixed (this is planed to be fixed in on our sprints, right?)

File: tests/e2e/backup_restore_suite_test.go [hunk 2]
```diff
@@ -176,7 +176,7 @@ func runApplicationBackupAndRestore(brCase ApplicationBackupRestoreCase, expecte
 	}
 }
 
-func runBackup(brCase BackupRestoreCase, backupName string) bool {
+func runBackup(brCase BackupRestoreCase, backupName string, expectedErr error) bool {
```
[the remark this answers is not in the document]
Comment (mrnold) [reply tests/e2e/backup_restore_suite_test.go:179]: I guess technically not, since I haven't added any tests with errors in the backup stage. But it looked weird to have it all plumbed through runRestore and not runBackup.

**Comment by weshayutin** [comment]
let's revive this work asap. Thank you!!

File: tests/e2e/backup_restore_suite_test.go (hunk 1 above)
Comment (mrnold) [reply tests/e2e/backup_restore_suite_test.go]: Okay, I tried this out in the latest push. The changes are definitely more contained now.

**Review by mateusoliveira43 — DISMISSED** [review]
changes seems fine, CI is happy with it

Would just increase a timeout for VM terminating, saw it timeout in my PR as well

**Review by mateusoliveira43 — APPROVED** [review]

File: tests/e2e/virt_backup_restore_suite_test.go [hunk 3]
```diff
@@ -73,18 +92,26 @@ func runVmBackupAndRestore(brCase VmBackupRestoreCase, expectedErr error, update
 		time.Sleep(brCase.InitDelay)
 	}
 
+	// Check if this VM should be running or stopped for this test.
+	// Depend on pre-backup verification function to poll state.
+	if brCase.PowerState == "Stopped" {
+		log.Print("Stopping VM before backup as specified in test case.")
+		err = v.StopVm(brCase.Namespace, brCase.Name)
+		gomega.Expect(err).To(gomega.BeNil())
```
Comment (kaovilai) [inline tests/e2e/virt_backup_restore_suite_test.go:100]: Do we need any waits or are stops instant?

Comment (mrnold) [reply tests/e2e/virt_backup_restore_suite_test.go:100]: I have a wait as the pre-backup verification function in the test table. I forget why I thought that made sense.

Comment (kaovilai) [reply tests/e2e/virt_backup_restore_suite_test.go:100]: +1 makes sense to me... just didn't connect the dots that there could be waits outside this function.

**Review by kaovilai — APPROVED** [review]

**Review by shubham-pampattiwar — APPROVED** [review]

**Comment by mateusoliveira43** [comment]
do we need to cherry-pick this oadp-1.4?

**Comment by mrnold** [comment]
> do we need to cherry-pick this oadp-1.4?

Yes, I think this was the plan. It doesn't merge cleanly though, so I will do it without the bot.

File: tests/e2e/sample-applications/virtual-machines/cirros-test/cirros-test.yaml [hunk 4]
```diff
@@ -33,6 +33,10 @@ items:
               - disk:
                   bus: virtio
                 name: rootdisk
+            firmware:
+              bootloader:
+                efi:
+                  secureBoot: false
```
Comment (mpryc) [inline tests/e2e/sample-applications/virtual-machines/cirros-test/cirros-test.yaml:39]: Why is this needed?

File: tests/e2e/virt_backup_restore_suite_test.go [hunk 5]
```diff
@@ -188,3 +230,45 @@ var _ = ginkgov2.Describe("VM backup and restore tests", ginkgov2.Ordered, func(
...
+	if brCase.RestoreErr != nil {
+		gomega.Expect(succeeded).To(gomega.Equal(false))
+		gomega.Expect(err).NotTo(gomega.BeNil())
+		gomega.Expect(err.Error()).To(gomega.ContainSubstring(brCase.RestoreErr.Error()))
+	} else {
+		gomega.Expect(err).ToNot(gomega.HaveOccurred())
+		gomega.Expect(succeeded).To(gomega.Equal(true))
+	}
+
+	if nsRequiresResticDCWorkaround {
```
Comment (mpryc) [inline tests/e2e/virt_backup_restore_suite_test.go:264]: Is this still needed?

File: tests/e2e/virt_backup_restore_suite_test.go [hunk 6]
```diff
@@ -37,10 +39,27 @@ func getLatestCirrosImageURL() (string, error) {
 	return imageURL, nil
 }
 
+func vmPoweredOff(vmnamespace, vmname string) VerificationFunction {
+	return VerificationFunction(func(ocClient client.Client, namespace string) error {
+		isOff := func() bool {
+			status, err := lib.GetVmStatus(dynamicClientForSuiteRun, vmnamespace, vmname)
+			if err != nil {
+				log.Printf("Error getting VM status: %v", err)
```
Comment (mpryc) [inline tests/e2e/virt_backup_restore_suite_test.go:47]: should there be a return on error?

File: tests/e2e/virt_backup_restore_suite_test.go (hunk 5 above)
Comment (mrnold) [reply tests/e2e/virt_backup_restore_suite_test.go:264]: I think possibly not, but I'm not really sure. I just copied all of this from the original runRestore, since it is planned to be temporary.

File: tests/e2e/virt_backup_restore_suite_test.go (hunk 6 above)
Comment (mrnold) [reply tests/e2e/virt_backup_restore_suite_test.go:47]: I think this could include an error, but I don't know that it would meaningfully improve the Eventually. I guess there might be cases where it would quit faster than the timeout, but then it would also need to check for errors that it should be expected to ride through. I will see if I can think of something.

Comment (mpryc) [reply tests/e2e/virt_backup_restore_suite_test.go:47]: I just see some possible logic problem. Saying first that there is err getting status and then without return logging off vm status and returning "Stopped", for me if there was error getting status the return of the status should not be "Stopped" but something unknown
